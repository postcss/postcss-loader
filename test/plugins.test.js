import path from 'path';

import {
  compile,
  getCompiler,
  getErrors,
  getCodeFromBundle,
  getWarnings,
} from './helpers';

import myPostcssPlugin from './fixtures/plugin/plugin';

jest.setTimeout(30000);

describe('"plugins" option', () => {
  it('should work with "Array"', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: [
          'postcss-nested',
          ['postcss-short', { prefix: 'x' }],
          myPostcssPlugin,
          // Like:
          // `
          // import myPlugin from './path/to/plugin.mjs';
          //
          // const initPlugin = myPlugin();
          // `
          (root) => {
            root.walkDecls((decl) => {
              if (decl.value === 'red') {
                // eslint-disable-next-line no-param-reassign
                decl.value = 'rgba(255, 0, 0, 1.0)';
              }
            });
          },
          // Like:
          // `
          // import myPlugin from './path/to/plugin.mjs';
          // `
          {
            postcss: (root) => {
              root.walkDecls((decl) => {
                if (decl.value === 'green') {
                  // eslint-disable-next-line no-param-reassign
                  decl.value = 'rgba(0, 255, 0, 1.0)';
                }
              });
            },
          },
          require.resolve('./fixtures/plugin/other-plugin'),
          myPostcssPlugin({ color: 'white', alpha: 0 }),
          { 'postcss-short': { prefix: 'z' } },
        ],
      },
    });
    const stats = await compile(compiler);
    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Object"', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: {
          'postcss-import': {},
          'postcss-nested': {},
          'postcss-short': { prefix: 'x' },
          [require.resolve('./fixtures/plugin/other-plugin')]: {},
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with empty "Array"', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: [],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with empty "Object"', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: {},
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Array" and support disabling plugins from the configuration', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        config: path.resolve(__dirname, './fixtures/css/plugins.config.js'),
        plugins: [{ 'postcss-short': false }],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Object" and support disabling plugins from the configuration', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        config: path.resolve(__dirname, './fixtures/css/plugins.config.js'),
        plugins: {
          'postcss-short': false,
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Object" and only disabled plugins', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: {
          'postcss-import': false,
          'postcss-nested': false,
          'postcss-short': false,
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Array" and override the previous plugin options', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: [
          ['postcss-short', { prefix: 'x' }],
          ['postcss-short', { prefix: 'z' }],
        ],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Array", and config, and override the previous plugin options', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        config: path.resolve(__dirname, './fixtures/css/plugins.config.js'),
        plugins: [['postcss-short', { prefix: 'z' }]],
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Object" and override the previous plugin options', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: {
          'postcss-short': { prefix: 'x' },
          // eslint-disable-next-line no-dupe-keys
          'postcss-short': { prefix: 'z' },
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should work with "Object", and config, and override the previous plugin options', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        config: path.resolve(__dirname, './fixtures/css/plugins.config.js'),
        plugins: {
          'postcss-short': { prefix: 'z' },
        },
      },
    });
    const stats = await compile(compiler);

    const codeFromBundle = getCodeFromBundle('style.css', stats);

    expect(codeFromBundle.css).toMatchSnapshot('css');
    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats)).toMatchSnapshot('errors');
  });

  it('should throw an error on the unresolved plugin', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        plugins: ['postcss-unresolved'],
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });

  it('should work with "Array" and not throw an error on falsy plugin', async () => {
    const compiler = getCompiler('./css/index.js', {
      postcssOptions: {
        // eslint-disable-next-line no-undefined
        plugins: [undefined, null, '', 0],
      },
    });
    const stats = await compile(compiler);

    expect(getWarnings(stats)).toMatchSnapshot('warnings');
    expect(getErrors(stats, true)).toMatchSnapshot('errors');
  });
});
