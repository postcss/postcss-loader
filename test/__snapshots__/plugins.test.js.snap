// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`"plugins" option should throw an error on the unresolved plugin: errors 1`] = `
Array [
  "ModuleError: Module Error (from \`replaced original path\`):
Loading PostCSS Plugin failed: Cannot find module 'postcss-unresolved' from 'src/utils.js'",
]
`;

exports[`"plugins" option should throw an error on the unresolved plugin: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and not throw an error on falsy plugin: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and not throw an error on falsy plugin: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and override the previous plugin options: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Array" and override the previous plugin options: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and override the previous plugin options: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and support disabling plugins from the configuration: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Array" and support disabling plugins from the configuration: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Array" and support disabling plugins from the configuration: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Array", and config, and override the previous plugin options: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Array", and config, and override the previous plugin options: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Array", and config, and override the previous plugin options: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Array": css 1`] = `
"a {
  color: rgba(0, 0, 0, 1.0);
}

a {
  color: rgba(255, 0, 0, 1.0);
}

a {
  color: rgba(0, 255, 0, 1.0);
}

a {
  color: rgba(0, 0, 255, 1.0);
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  border-top-color: rgba(0, 0, 255, 1.0);
  border-right-color: rgba(0, 0, 255, 1.0);
  border-left-color: rgba(0, 0, 255, 1.0);
  background-color: #fafafa;
}

.phone_title {
    width: 500px;
  }

@media (max-width: 500px) {

.phone_title {
      width: auto
  }
    }

body.is_dark .phone_title {
      color: rgba(0, 0, 0, 0);
    }

.phone img {
    display: block;
  }
"
`;

exports[`"plugins" option should work with "Array": errors 1`] = `Array []`;

exports[`"plugins" option should work with "Array": warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and only disabled plugins: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Object" and only disabled plugins: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and only disabled plugins: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and override the previous plugin options: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Object" and override the previous plugin options: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and override the previous plugin options: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and support disabling plugins from the configuration: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Object" and support disabling plugins from the configuration: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Object" and support disabling plugins from the configuration: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Object", and config, and override the previous plugin options: css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  border-top-color: blue;
  border-right-color: blue;
  border-left-color: blue;
  background-color: #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with "Object", and config, and override the previous plugin options: errors 1`] = `Array []`;

exports[`"plugins" option should work with "Object", and config, and override the previous plugin options: warnings 1`] = `Array []`;

exports[`"plugins" option should work with "Object": css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: rgba(0, 0, 255, 1.0);
}

.class {
  border-top-color: rgba(0, 0, 255, 1.0);
  border-right-color: rgba(0, 0, 255, 1.0);
  border-left-color: rgba(0, 0, 255, 1.0);
  background-color: #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone_title {
    width: 500px;
  }

@media (max-width: 500px) {

.phone_title {
      width: auto
  }
    }

body.is_dark .phone_title {
      color: white;
    }

.phone img {
    display: block;
  }
"
`;

exports[`"plugins" option should work with "Object": errors 1`] = `Array []`;

exports[`"plugins" option should work with "Object": warnings 1`] = `Array []`;

exports[`"plugins" option should work with empty "Array": css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with empty "Array": errors 1`] = `Array []`;

exports[`"plugins" option should work with empty "Array": warnings 1`] = `Array []`;

exports[`"plugins" option should work with empty "Object": css 1`] = `
"a {
  color: black;
}

a {
  color: red;
}

a {
  color: green;
}

a {
  color: blue;
}

.class {
  -x-border-color: blue blue *;
  -x-color: * #fafafa;
}

.class-foo {
  -z-border-color: blue blue *;
  -z-color: * #fafafa;
}

.phone {
  &_title {
    width: 500px;

    @media (max-width: 500px) {
      width: auto;
    }

    body.is_dark & {
      color: white;
    }
  }

  img {
    display: block;
  }
}
"
`;

exports[`"plugins" option should work with empty "Object": errors 1`] = `Array []`;

exports[`"plugins" option should work with empty "Object": warnings 1`] = `Array []`;
